File: src/main/java/com/challenge/interview/calendar/services/AvailabilityService.java
package com.challenge.interview.calendar.services;

import com.challenge.interview.calendar.model.AvailabilityEntity;
import com.challenge.interview.calendar.model.Role;
import com.challenge.interview.calendar.model.TimeInterval;
import com.challenge.interview.calendar.repository.AvailabilityRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class AvailabilityService {

    private AvailabilityRepository repository;

    @Autowired
    public AvailabilityService (AvailabilityRepository repository) {
        this.repository = repository;
    }

    public List<TimeInterval> getAvailabilities(String candidateId, List<String> interviewerIds) {
        List<AvailabilityEntity> candidateAvailabilities = repository.findByCalendarIdAndRole(candidateId, Role.CANDIDATE.toString());
        List<List<AvailabilityEntity>> interviewersAvailabilities = interviewerIds.stream()
                .map((interviewerId) -> repository.findByCalendarIdAndRole(interviewerId, Role.INTERVIEWER.toString()))
                .collect(Collectors.toList());

        final List<TimeInterval>[] intersectedIntervals = (List<TimeInterval>[]) new Object[1];
        intersectedIntervals[0] = candidateAvailabilities.stream()
                .map((availability) -> availability.toTimeInterval())
                .collect(Collectors.toList());

        interviewersAvailabilities.forEach((interviewerAvailability) -> {
            List<TimeInterval> interviewerIntervals = interviewerAvailability.stream()
                    .map((availability) -> availability.toTimeInterval())
                    .collect(Collectors.toList());
            intersectedIntervals[0] = intersect(intersectedIntervals[0], interviewerIntervals);
        });

        return intersectedIntervals[0];
    }

    private List<TimeInterval> intersect(List<TimeInterval> intersectedIntervals, List<TimeInterval> interviewerIntervals) {
        ArrayList<TimeInterval> list = new ArrayList<>();
        interviewerIntervals.forEach((interval) -> {
            interviewerIntervals.forEach((interviewerTimeInterval) -> {
                TimeInterval overlapInterval = interval.overlap(interviewerTimeInterval);
                if (overlapInterval != null)
                    list.add(overlapInterval);
            });
        });
        return list;
    }
}

File: src/main/kotlin/com/challenge/interview/calendar/services/AvailabilityService.kt
package com.challenge.interview.calendar.services

import com.challenge.interview.calendar.model.Role
import com.challenge.interview.calendar.model.TimeInterval
import com.challenge.interview.calendar.repository.AvailabilityRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class AvailabilityService
    @Autowired
    constructor (private val repository: AvailabilityRepository) {

    private val log = LoggerFactory.getLogger(this.javaClass.name)

    fun getAvailabilities(candidateId: String, interviewerIds: List<String>): List<TimeInterval> {
        val candidateAvailabilities = repository.findByCalendarIdAndRole(candidateId, Role.CANDIDATE.toString());
        val interviewerAvailabilities = interviewerIds.map {
            repository.findByCalendarIdAndRole(it, Role.INTERVIEWER.toString())
        }

        //val candidateTimeIntervals = candidateAvailabilities.map { it.toTimeInterval() }
        var intersectedIntervals = candidateAvailabilities.map { it.toTimeInterval() }
        interviewerAvailabilities.forEach {
            val interviewerIntervals = it.map { it.toTimeInterval() }
            intersectedIntervals = intersect(intersectedIntervals, interviewerIntervals)
        }

        return intersectedIntervals
    }

    private fun intersect(intersectedIntervals: List<TimeInterval>, interviewerIntervals: List<TimeInterval>): List<TimeInterval> {
        val list = mutableListOf<TimeInterval>()
        interviewerIntervals.forEach{
            val interval = it
            interviewerIntervals.forEach{
                val overlapInterval = interval.overlap(it)
                if (overlapInterval != null)
                    list.add(overlapInterval)
            }
        }
        return list
    }
}